From 2bf8bcc438c52eae244f0920cc62b2d1707bd0a3 Mon Sep 17 00:00:00 2001
From: Davin McCall <davmac@davmac.org>
Date: Tue, 7 Nov 2023 07:23:49 +1000
Subject: [PATCH] control: process all available input buffer

Process input buffer in a loop until it has all been consumed or we only
have a partial packet.
---
 src/control.cc | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/src/control.cc b/src/control.cc
index 63933c9d..e8ebca40 100644
--- a/src/control.cc
+++ b/src/control.cc
@@ -15,7 +15,7 @@
 // Control protocol versions:
 // 1 - dinit 0.16 and prior
 // 2 - dinit 0.17 (adds SETTRIGGER, CATLOG, SIGNAL)
-// 3 - (unreleased) (adds QUERYSERVICEDSCDIR)
+// 3 - dinit 0.17.1 (adds QUERYSERVICEDSCDIR)
 
 // common communication datatypes
 using namespace dinit_cptypes;
@@ -1345,7 +1345,7 @@ bool control_conn_t::data_ready() noexcept
     }
     
     // complete packet?
-    if (rbuf.get_length() >= chklen) {
+    while (rbuf.get_length() >= chklen) {
         try {
             return !process_packet();
         }
@@ -1353,8 +1353,11 @@ bool control_conn_t::data_ready() noexcept
             do_oom_close();
             return false;
         }
+
+        chklen = std::max(chklen, 1u);
     }
-    else if (rbuf.get_length() == rbuf.get_size()) {
+
+    if (rbuf.get_length() == rbuf.get_size()) {
         // Too big packet
         log(loglevel_t::WARN, "Received too-large control packet; dropping connection");
         bad_conn_close = true;
